neo4j/bookmarks.go

```diff
@@ -0,0 +1,57 @@
+package neo4j
+
+/*
+Bookmarks is a holder for server-side bookmarks which are used for causally chaining sessions.
+
+See also CombineBookmarks.
+
+Note:
+Will be changed from being a type alias to being a struct in 6.0. Please use BookmarksFromRawValues for construction
+from raw values and BookmarksToRawValues for accessing the raw values.
+*/
+type Bookmarks = []string
+
+/*
+CombineBookmarks is a helper method to combine []Bookmarks into a single Bookmarks instance.
+
+Let s1, s2, s3 be Session interfaces. You can easily causally chain the sessions like so:
+
+	s4 := driver.NewSession(neo4j.SessionConfig{
+		Bookmarks: neo4j.CombineBookmarks(s1.LastBookmarks(), s2.LastBookmarks(), s3.LastBookmarks()),
+	})
+
+The server will then make sure to execute all transactions in s4 after any that were already executed in s1, s2, or s3
+at the time of calling LastBookmarks.
+*/
+func CombineBookmarks(bookmarks... Bookmarks) Bookmarks  {
+	var lenSum int
+	for _, b := range bookmarks {
+		lenSum += len(b)
+	}
+	res := make([]string, lenSum)
+	var i int
+	for _, b := range bookmarks {
+		i += copy(res[i:], b)
+	}
+	return res
+}
+
+/*
+BookmarksToRawValues exposes the raw server-side bookmarks.
+
+You should not need to use this method unless you want to serialize bookmarks.
+See Session.LastBookmarks and CombineBookmarks for alternatives.
+ */
+func BookmarksToRawValues(bookmarks Bookmarks) []string {
```

Comment:
Ideally, I'd like this to be a receiver function of `Bookmarks`. But since it's a type alias, it can't be one. Maybe deprecate this in 6.0 when `Bookmarks` becomes a real struct and introduce an alternative for this function that is a receiver, then remove this function in 7.0? I hate the massively long scope of this. Any better ideas?

Reply:
since this is only exposed for tests (integration tests and testkit): can we move (and duplicate if needed) this function?

Reply:
It's not for tests only. The the docs say, if you want to serialize bookmarks (e.g. to achieve causal consistency between different processes or machines), you need a way to dump/load bookmarks to/fro something easily serializable (e.g. `[]string`).